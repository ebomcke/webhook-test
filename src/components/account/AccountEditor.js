import React, { Component } from 'react';
import { compose } from 'recompose';
import { withFirestore } from 'react-firestore';
import { withFirebase } from '../../contexts/Firebase';
import { withAccount } from '../../contexts/Account';

class AccountEditor extends Component {
  constructor(props) {
    super(props);
    let { account } = this.props;
    this.state = account;
  }

  handleInputChange = ({ event, name, data }) => {
    const target = event.target;
    let value;
    if (data) {
      value = data.value;
    } else {
      value = target.type === 'checkbox' ? target.checked : target.value;
    }
    if (!name) {
      name = target.name;
    }

    this.setState({
      [name]: value,
    });
  }

  save = async () => {
    const { firebase, firestore } = this.props;
    if (!this.state.confirmed) {
      this.state.confirmed = firebase.firestore.Timestamp.now();
    }
    await firestore
      .collection('accounts')
      .doc(this.state.id)
      .set(this.state);
    return this.state.id;
  }

  render() {
    const { children } = this.props;
    return children({
      account: this.state,
      handleInputChange: this.handleInputChange,
      save: this.save,
    });
  }
}

export default compose(
  withFirebase,
  withFirestore,
  withAccount
)(AccountEditor);
