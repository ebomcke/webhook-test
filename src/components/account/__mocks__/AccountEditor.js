const mockAccountEditor = {
    account: {
        name: 'Unit Test',
        organisationName: 'unit-test-org-name'
    },
    onOrganisationNameChange: jest.fn(),
    save: jest.fn(() => Promise.resolve())
}

export { mockAccountEditor };