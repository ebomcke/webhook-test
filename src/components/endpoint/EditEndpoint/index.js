import React from 'react';
import { Form, Header } from 'semantic-ui-react';
import Url from './Url';
import Response from './Response';

const EditEndpoint = ({ endpointEditor }) => (
  <div>
    <Header as="h1">Edit endpoint</Header>
    <Url endpointEditor={endpointEditor} disabled={true} />
    <p />
    <Response
      renderMessage={false}
      endpointEditor={endpointEditor}
      nextButton={() => (
        <Form.Button name="endpoint-editor-save" onClick={endpointEditor.save}>
          Save
        </Form.Button>
      )}
    />
  </div>
);

export const EditEndpointUrl = Url;
export const EditEndpointResponse = Response;

export default EditEndpoint;
