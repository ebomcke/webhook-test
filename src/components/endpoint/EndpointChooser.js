import React from 'react';
import { compose } from 'recompose';
import { withAccount } from '../../contexts/Account';
import { Dropdown } from 'semantic-ui-react';
import { withEndpoints } from '../../contexts/Endpoints';

const EndpointChooser = ({ account, endpoints, onChange }) => (
  <Dropdown
    selection
    search
    options={endpoints.map(endpoint => ({
      text: `/${account.organisationName}/${endpoint.path}`,
      value: endpoint.id,
    }))}
    onChange={(event, data) =>
      onChange(endpoints.find(e => e.id === data.value))
    }
  />
);
export default compose(
  withAccount,
  withEndpoints,
)(EndpointChooser);
