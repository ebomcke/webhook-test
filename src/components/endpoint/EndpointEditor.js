import React, { Component } from 'react';
import { compose } from 'recompose';
import { withFirestore } from 'react-firestore';
import { withFirebase } from '../../contexts/Firebase';
import { withAccount } from '../../contexts/Account';
import { withNavigation } from '../../contexts/Navigation';

class EndpointEditor extends Component {
  constructor(props) {
    super(props);
    let { endpoint } = this.props;
    if (!endpoint) {
      endpoint = this.generateNewEndpoint();
    }
    this.state = {
      id: endpoint.id,
      path: endpoint.path,
      defaultResponseBody: endpoint.defaultResponse.body,
      defaultResponseStatusCode: endpoint.defaultResponse.statusCode,
    };
  }

  generateNewEndpoint = () => ({
    path: '',
    defaultResponse: {
      statusCode: 200,
      body: JSON.stringify(
        {
          status: 'OK',
        },
        null,
        2,
      ),
    },
  });

  onStatusCodeChange = (statusCode) => {
    this.setState({
      defaultResponseStatusCode: statusCode,
    });
  }

  onUrlChange = (url) => {
    url = url
      .toLowerCase()
      .replace(/ /gi, '')
      .replace(/^\//, '');
    this.setState({
      path: url,
    });
  }

  onResponseBodyChange = (value) => {
    try {
      JSON.parse(value);
      this.setState({
        defaultResponseBody: value,
      });
    } catch (e) {}
  }

  save = async () => {
    const { firebase, firestore, account, navigate } = this.props;
    let endpoint = {
      account: firestore.collection('accounts').doc(account.id),
      path: this.state.path,
      lastActive: firebase.firestore.Timestamp.now(),
      defaultResponse: {
        statusCode: this.state.defaultResponseStatusCode,
        body: this.state.defaultResponseBody,
      },
    };
    if (this.state.id) {
      endpoint.id = this.state.id;
    }
    if (endpoint.id) {
      await firestore
        .collection('endpoints')
        .doc(endpoint.id)
        .set(endpoint);
    } else {
      await firestore.collection('endpoints').add(endpoint);
    }
    navigate({
      name: 'Endpoints',
    });
  }

  render() {
    const { children } = this.props;
    return children({
      endpoint: this.state,
      onResponseBodyChange: this.onResponseBodyChange,
      onUrlChange: this.onUrlChange,
      onStatusCodeChange: this.onStatusCodeChange,
      save: this.save,
    });
  }
}

export default compose(
  withFirebase,
  withFirestore,
  withNavigation,
  withAccount,
)(EndpointEditor);
