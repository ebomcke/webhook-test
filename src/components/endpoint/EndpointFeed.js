import React from 'react';
import { Feed, Button } from 'semantic-ui-react';
import EndpointItem from './EndpointItem';
import { compose } from 'recompose';
import { withNavigation } from '../../contexts/Navigation';
import { withEndpoints } from '../../contexts/Endpoints';

const EndpointFeed = ({ endpoints, navigate }) => (
  <Feed>
    <Button
      size="big"
      content="New endpoint"
      icon="plus"
      onClick={() =>
        navigate({
          name: 'NewEndpoint',
        })
      }
    />
    {endpoints.map(endpoint => (
      <EndpointItem key={endpoint.id} endpoint={endpoint} />
    ))}
  </Feed>
);

export default compose(withNavigation, withEndpoints)(EndpointFeed);
