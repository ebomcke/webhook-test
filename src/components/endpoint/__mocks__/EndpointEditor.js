const mockEndpointEditor = {
    endpoint: {
        path: 'unit-test-path',
        defaultResponse: {
            statusCode: 200,
            body: JSON.stringify({status:'OK'})
        }
    },
    onResponseBodyChange: jest.fn(),
    onUrlChange: jest.fn(),
    onStatusCodeChange: jest.fn(),
    save: jest.fn(() => Promise.resolve())
}

export { mockEndpointEditor };