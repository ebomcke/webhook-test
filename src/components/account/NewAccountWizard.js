import React, { Component } from 'react';
import { Segment, Header, Icon, Message, Form } from 'semantic-ui-react';
import { compose } from 'recompose';
import { withNavigation } from '../../contexts/Navigation';
import Wizard from '../base/Wizard';
import { EditEndpointUrl } from '../endpoint/EditEndpoint';

class NewAccountWizard extends Component {
  constructor(props) {
    super(props);
  }

  save = async () => {
    const { accountEditor, endpointEditor, navigate } = this.props;
    await accountEditor.save();
    await endpointEditor.save();
    navigate({
      name: 'Home',
    });
  };

  render() {
    const { accountEditor, endpointEditor } = this.props;
    const { account, onOrganisationNameChange } = accountEditor;
    const { endpoint } = endpointEditor;
    const steps = [
      {
        key: 'organisation',
        title: 'Organisation name',
        description: 'Choose a name for your organisation',
        nextDisabled: () => !account.organisationName,
        message: () => (
          <Message icon>
            <Icon name="help" />
            <Message.Content>
              <Message.Header>Why do I need this?</Message.Header>
              <p>
                The organisation name that you choose will be used to create a
                unique URL prefix where you can send your webhooks.
              </p>
            </Message.Content>
          </Message>
        ),
        render: nextButton => (
          <Form>
            <Form.Field>
              <Form.Input
                label="Organisation name"
                fluid
                placeholder="mybrilliantcompany"
                name="organisationName"
                value={account.organisationName}
                onChange={(event, data) => onOrganisationNameChange(data.value)}
              />
            </Form.Field>
            {nextButton()}
          </Form>
        ),
      },
      {
        key: 'url',
        icon: 'at',
        title: 'URL',
        description: 'Create your first URL',
        nextDisabled: () => !endpoint.path,
        render: nextButton => (
          <EditEndpointUrl
            endpointEditor={endpointEditor}
            nextButton={nextButton}
          />
        ),
      },
    ];
    return (
      <Segment className="full-height">
        <Header as="h1">Good to see you, {account.name}</Header>
        <Wizard steps={steps} onComplete={this.save} />
      </Segment>
    );
  }
}

export default compose(withNavigation)(NewAccountWizard);
