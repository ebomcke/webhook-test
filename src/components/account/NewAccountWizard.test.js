import React from 'react';
import { create } from 'react-test-renderer';
import NewAccountWizard from './NewAccountWizard';

jest.mock('./AccountEditor');
import { mockAccountEditor } from './AccountEditor';
jest.mock('../endpoint/EndpointEditor');
import { mockEndpointEditor } from '../endpoint/EndpointEditor';

describe('Rendering tests', () => {
  test('Step 1 - organisation name', () => {
    const newAccountWizard = create(
      <NewAccountWizard
        accountEditor={mockAccountEditor}
        endpointEditor={mockEndpointEditor}
      />,
    );
    expect(newAccountWizard.toJSON()).toMatchSnapshot();
  });

  test('Step 2 - endpoint url', () => {
    const newAccountWizard = create(
      <NewAccountWizard
        accountEditor={mockAccountEditor}
        endpointEditor={mockEndpointEditor}
      />,
    );
    const nextButton = newAccountWizard.root.findByProps({
      wizardIdentifier: 'next-done-button',
    });
    nextButton.props.onClick();
    expect(newAccountWizard.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  afterEach(() => {
    mockAccountEditor.save.mockClear();
    mockEndpointEditor.save.mockClear();
  });
  test('Save triggers account & endpoint save and navigates home', async () => {
    const navigate = jest.fn();
    const newAccountWizard = create(
      <NewAccountWizard
        accountEditor={mockAccountEditor}
        endpointEditor={mockEndpointEditor}
        navigate={navigate}
      />,
    );
    const nextButton = newAccountWizard.root.findByProps({
      wizardIdentifier: 'next-done-button',
    });
    nextButton.props.onClick();
    const doneButton = newAccountWizard.root.findByProps({
      wizardIdentifier: 'next-done-button',
    });
    await doneButton.props.onClick();
    expect(mockAccountEditor.save.mock.calls.length).toBe(1);
    expect(mockEndpointEditor.save.mock.calls.length).toBe(1);
    expect(navigate.mock.calls.length).toBe(1);
    expect(navigate.mock.calls[0][0]).toEqual({
      name: 'Home',
    });
  });
});
