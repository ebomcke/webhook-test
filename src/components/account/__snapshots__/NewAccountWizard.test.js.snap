// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests Step 1 - organisation name 1`] = `
<div
  className="ui segment full-height"
>
  <h1
    className="ui header"
  >
    Good to see you, 
    Unit Test
  </h1>
  <div>
    <div
      className="ui fluid steps"
    >
      <div
        className="active step"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="title"
          >
            Organisation name
          </div>
          <div
            className="description"
          >
            Choose a name for your organisation
          </div>
        </div>
      </div>
      <div
        className="disabled step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="at icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            URL
          </div>
          <div
            className="description"
          >
            Create your first URL
          </div>
        </div>
      </div>
    </div>
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <div
          className="field"
        >
          <label>
            Organisation name
          </label>
          <div
            className="ui fluid input"
          >
            <input
              name="organisationName"
              onChange={[Function]}
              placeholder="mybrilliantcompany"
              type="text"
              value="unit-test-org-name"
            />
          </div>
        </div>
      </div>
      <div
        className="field"
      >
        <button
          className="ui button"
          onClick={[Function]}
          wizardIdentifier="next-done-button"
        >
          Next
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Rendering tests Step 2 - endpoint url 1`] = `
<div
  className="ui segment full-height"
>
  <h1
    className="ui header"
  >
    Good to see you, 
    Unit Test
  </h1>
  <div>
    <div
      className="ui fluid steps"
    >
      <div
        className="step"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="title"
          >
            Organisation name
          </div>
          <div
            className="description"
          >
            Choose a name for your organisation
          </div>
        </div>
      </div>
      <div
        className="active step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="at icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            URL
          </div>
          <div
            className="description"
          >
            Create your first URL
          </div>
        </div>
      </div>
    </div>
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="required field"
      >
        <label>
          URL
        </label>
        <div
          className="ui labeled input"
        >
          <div
            className="ui label label"
            onClick={[Function]}
          >
            https://webhook-tester-1547141127436.appspot.com/webhook/unit-test-org/
          </div>
          <input
            name="path"
            onChange={[Function]}
            placeholder="my/custom/url"
            type="text"
            value="unit-test-path"
          />
        </div>
      </div>
      <div
        className="field"
      >
        <button
          className="ui button"
          onClick={[Function]}
          wizardIdentifier="next-done-button"
        >
          Done
        </button>
      </div>
    </form>
  </div>
</div>
`;
